### PR TITLE
Add `bayes_factor` Function for Computations and Test Cases (#45)

### PR DESCRIPTION
**Pull request by PiyushPanwarFST**
This PR addresses issue [#45](https://github.com/arviz-devs/arviz-stats/issues/45) by refactoring and enhancing the Bayes Factor computation functionality. Key updates include:

## New Functionality

Added bayes_factor in psense.py for computing the Bayes Factor statistic, with optional plotting support.
Decoupled computation logic `az.bayes_factor` from visualisation `az.plot_bf` to ensure better modularity.

## Test Coverage

Added comprehensive tests in `test_psense.py` to validate functionality and cover edge cases.

## Code Quality

Documented the function using Numpy-style docstrings.
Ensured proper formatting with black and verified code quality using pylint and lint.sh.

## Benefits

Clear separation of concerns between computation `az.bayes_factor` and visualisation `az.plot_bf`.
Enhanced focus on statistical computations in arviz-stats.
Simplified dependencies for users needing computation-only logic.

## Additional Suggestion

I propose adding _kde_linear functionality to the density.py file, as implemented in the main ArviZ repository. This addition would further improve the modularity and functionality of  `arviz-stats`. I plan to raise a separate issue for this enhancement.

----
📚 Documentation preview 📚: https://arviz-stats--46.org.readthedocs.build/en/46/

### CODE REVIEW

**Comment by aloctavodia** [comment]
bayes_factor should not be in psense. Those are two unrelated functions. Don't use gaussian_kde from SciPy. ArviZ already have a kde functionality, which is much better than the one from scipy.

Also not clear why are you removing functionality from the density module.

**Comment by PiyushPanwarFST** [comment]
KDE Implementation:
Initially, I used KDE since the inference_data was not available in arviz-stats, that's why I use scipy But if needed, I can switch back to using KDE again.

Density Module Context:
So initially, the density module only have circular_kde not kde_linear I added the functionality in density module , but when I used scipy, this new functionality didn’t really come into use that' why I removing it again.

Question I want to ask:
1. Sir could you please let me know where exactly I should add the bayes_factor function ?? 
2. As we are using the kde functionality so do we have to add the inference_data module in arviz-stats ??

**Comment by aloctavodia** [comment]
I recomend you first get familiar with the library and how is organized. Check the documentation to see how it is used and then inspect the repo to see how functions are internally organized. 

When I said that you could follow psense implementation I ment that bayes_factor could be implemented in a similar way not in the same file. The main reason is that bayes factor needs access to prior and posterior groups. Compare this to how ess or rhat are implemented (as accesors) and apply to specific groups. 

We no longer use InferenceData, see arviz-base for the equivalent structures, like DataTree, DataArray, and Dataset. You can use functionality from arviz-base if needed.

**Comment by PiyushPanwarFST** [comment]
I think the bayes_factor function takes a DataArray as input, so we should add it to dataarray.py because all the functions that take DataArray as input are in this file, right?

And as I reviewed the bayes_factor function in the main repo, they also use kde_linear. So, I think we should also use the kde_linear ??

**Comment by aloctavodia** [comment]
As a fist step add all code into a file called bayes_factor.py, the file should be at the same level as psense.py. For discrete posterior/prior you can compute the BF as:

```python
ref_val = 0
idata = azb.from_dict(
    {"posterior":{"a": np.random.poisson(3.2, (2, 5000))}, "prior":{"a": np.random.poisson(3.2, (2, 5000))}}
)
posterior_at_ref_val = (idata.posterior["a"].azstats.kde() == ref_val).mean()
prior_at_ref_val = (idata.prior["a"].azstats.kde() == ref_val).mean()
bf_10 = prior_at_ref_val / posterior_at_ref_val
```

if data is continuous, this is slightly more complex as we need to do an interpolation.

**Comment by PiyushPanwarFST** [comment]
Screenshot 2025-01-14 at 9 40 34 PM


Sir although I initialise the kde_linear function in the Base_Array class in array.py file but still I am getting the error ??
How they are dependent on the test_kde_is_normalized or test_hdi_multimodal_continous

**Comment by aloctavodia** [comment]
don't use that function, follow my example.

**Comment by PiyushPanwarFST** [comment]
Sir, following the example you provided, I have written the `bayes_factor` function in the `bayes_factor.py` file, and all the tests cases which I written for the `bayes_factor` function have passed successfully. However, I am still getting the same 10 errors, which state that the kde_linear function is not defined in the `BaseArray` class.

To resolve this, I created a new virtual environment and reinstalled all the dependencies, but still I got the same errors

Could you please guide me on how to resolve this issue?



def bayes_factor(idata, var_name, ref_val=0, return_ref_vals=False, prior= None):
    
    """
    Approximated Bayes Factor for comparing hypotheses of two nested models, 
    using KDE for density estimation.

    Parameters
    ----------
    idata : InferenceData
        Object containing posterior and prior data.
    var_name : str
        Name of the variable to test.
    ref_val : int or float, default 0
        Reference (point-null) value for Bayes factor estimation.
    return_ref_vals : bool, default False
        If True, also return the values of prior and posterior densities at the reference value.

    Returns
    -------
    dict
        A dictionary with Bayes Factor values: BF10 (H1/H0 ratio) and BF01 (H0/H1 ratio).
    """

    # Extract posterior samples for the specified variable
    posterior = idata.posterior[var_name].values.flatten()
    prior = idata.prior[var_name].values.flatten()

    # Remove NaN values for both posterior and prior
    posterior = posterior[np.isfinite(posterior)]
    prior = prior[np.isfinite(prior)]
    

    # Validate ref_val
    if not isinstance(ref_val, (int, float)):
        raise ValueError("The reference value (ref_val) must be a numerical value (int or float).")
    
    # Warn if the reference value is outside the posterior range
    if ref_val > posterior.max() or ref_val < posterior.min():
        warnings.warn(
            "The reference value is outside the posterior range. "
            "This results in infinite support for H1, which may overstate evidence."
        )

    # Compute KDE for posterior and prior
    posterior_grid, posterior_pdf = azb.kde(x=posterior, grid_len=512, circular=False)
    prior_grid, prior_pdf = azb.kde(x=prior, grid_len=512, circular=False)


    # Interpolate the densities at the reference value
    posterior_at_ref_val = np.interp(ref_val, posterior_grid, posterior_pdf)
    prior_at_ref_val = np.interp(ref_val, prior_grid, prior_pdf)

    # Compute Bayes Factor
    bf_10 = prior_at_ref_val / posterior_at_ref_val
    bf = {"BF10": bf_10, "BF01": 1 / bf_10}

    # Optionally return prior and posterior values at the reference value
    if return_ref_vals:
        return (bf, {"prior": prior_at_ref_val, "posterior": posterior_at_ref_val})
    else:
        return bf

**Comment by aloctavodia** [comment]
Please add code into commits, no comments, unless there are short snippets. Also if you want to share error share the traceback. I think we don't have functions yet in arviz-stats. Only accessors. So azb.kde is not going to work. Moreover, azb seems to be an alias of arviz-base, but arviz-base does not have KDE. Density estimation lives in arviz-stats.

**Comment by PiyushPanwarFST** [comment]
Sir, as you mentioned that azb.kde is not going to work, I have modified the code for Bayes_factor. Could you please review it and let me know if there are any mistakes or areas for improvement?

**Comment by aloctavodia** [comment]
The last commits includes changes to a lot of files, most of them unrelated to this PR, please remove non necessary changes/files

**Comment by PiyushPanwarFST** [comment]
Is something I need to change in the codebase to resolve the error ??

**Comment by aloctavodia** [comment]
I still see changes in 25 files, that's a lot. Please remove unnecessary changes/files and push the changes related to this PR and nothing else.

**Comment by PiyushPanwarFST** [comment]
Sir I only added two new files and otherwise I have't touch the codebase and the changes in 6 files are just line deletion and addition of same line, otherwise anything else

File: src/arviz_stats/base/array.py [hunk 1]
```diff
@@ -333,4 +333,4 @@ def kde(self, ary, axes=-1, circular=False, grid_len=512, **kwargs):
         )
 
 
-array_stats = BaseArray()
+array_stats = BaseArray()
```
Comment (aloctavodia) [inline src/arviz_stats/base/array.py:336]:
```suggestion
array_stats = BaseArray()

```

File: src/arviz_stats/base/density.py [hunk 2]
```diff
@@ -781,4 +781,4 @@ def _find_hdi_contours(self, density, hdi_probs):
             except IndexError:
                 contours[idx] = sorted_density[0]
 
-        return contours
+        return contours
```
Comment (aloctavodia) [inline src/arviz_stats/base/density.py:784]:
```suggestion
        return contours
        
```

File: src/arviz_stats/psense.py [hunk 3]
```diff
@@ -305,4 +305,4 @@ def _get_power_scale_weights(
     upper_w = np.exp(group_draws.azstats.power_scale_lw(alpha=alphas[1], dims=sample_dims))
     upper_w = upper_w / upper_w.sum(sample_dims)
 
-    return lower_w, upper_w
+    return lower_w, upper_w
```
Comment (aloctavodia) [inline src/arviz_stats/psense.py:308]:
```suggestion
    return lower_w, upper_w
    
```

File: tests/test_psense.py [hunk 4]
```diff
@@ -32,4 +32,4 @@ def test_psense_summary():
     assert all(psense_df["diagnosis"] == "prior-data conflict")
 
     psense_df = psense_summary(uni_dt, threshold=1)
-    assert all(psense_df["diagnosis"] == "✓")
+    assert all(psense_df["diagnosis"] == "✓")
```
Comment (aloctavodia) [inline tests/test_psense.py:35]:
```suggestion
    assert all(psense_df["diagnosis"] == "✓")
    
```

File: src/arviz_stats/base/density.py [hunk 5]
```diff
@@ -782,3 +782,4 @@ def _find_hdi_contours(self, density, hdi_probs):
                 contours[idx] = sorted_density[0]
 
         return contours
+
```
Comment (aloctavodia) [inline src/arviz_stats/base/density.py:785]:
```suggestion
```

File: src/arviz_stats/base/density.py (hunk 2 above)
Comment (aloctavodia) [inline src/arviz_stats/base/density.py:784]:
```suggestion
        return contours
```

File: src/arviz_stats/base/density.py [hunk 6]
```diff
@@ -782,3 +782,7 @@ def _find_hdi_contours(self, density, hdi_probs):
                 contours[idx] = sorted_density[0]
 
         return contours
+<<<<<<< HEAD
+    
+=======
+>>>>>>> origin/feature
```
Comment (aloctavodia) [inline src/arviz_stats/base/density.py:788]:
```suggestion
```

File: tests/test_psense.py [hunk 7]
```diff
@@ -33,3 +33,7 @@ def test_psense_summary():
 
     psense_df = psense_summary(uni_dt, threshold=1)
     assert all(psense_df["diagnosis"] == "✓")
+<<<<<<< HEAD
+    
+=======
+>>>>>>> origin/feature
```
Comment (aloctavodia) [inline tests/test_psense.py:39]:
```suggestion
```

**Review by aloctavodia — CHANGES_REQUESTED** [review]

File: src/arviz_stats/bayes_factor.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+    if ref_val > posterior.max() or ref_val < posterior.min():
+        warnings.warn(
+            "The reference value is outside the posterior range. "
+            "This results in infinite support for H1, which may overstate evidence."
+        )
+    if posterior.dtype.kind == "f":
+        density_instance = _DensityBase()
+        posterior_grid, posterior_pdf, _ = density_instance._kde(x=posterior, grid_len=512, circular=False)
+        prior_grid, prior_pdf, _ = density_instance._kde(x=prior, grid_len=512, circular=False)
```
Comment (aloctavodia) [inline src/arviz_stats/bayes_factor.py:38]: @OriolAbril what do you think is the best way to implement this function? In particular this part?

**Comment by PiyushPanwarFST** [comment]
Sir what is the status of my PR , is there any need to modify the code or something else ??

**Comment by PiyushPanwarFST** [comment]
Sir is there any update regarding the PR ??

**Comment by aloctavodia** [comment]
I appreciate the enthusiasm! However, you must understand the nature of open-source projects and contributions.  People work on open-source projects mostly in their free time. Asking twice the same thing two days in a row, and even more during a weekend is not the way to behave. Sometimes developer work on weekends, but is their choice, don't be pushy. 

Notice that I have asked another developer for his feedback, he has not answered yet, maybe because he is enjoying the weekend, maybe because he is working on other stuff, who knows! Learn to be patient.

I did not have the time to check this from last week, because I have been busy with other tasks. So please wait for the status of this PR to change. It is ok to ask from time to time, but I would wait at least one week. If you have free time and want to help with this project (help is always welcome) you could check for other issues, read the documentation, read the codebase,  work on a different PR, etc.

**Comment by PiyushPanwarFST** [comment]
I hope you're doing well. First, I want to sincerely apologize for the mistake. I understand that open-source projects are driven by contributors’ personal time and schedules, and I appreciate all the effort you and others put into this. I will ensure to be more patient in the future and avoid being pushy.

I have quick question regarding the implementation of Bayes Factor plotting functionality in arviz-plots. While the Bayes Factor computation has been completed in the main ArviZ repo and arviz-stats, I noticed that the plotting functionality for Bayes Factor seems to be pending in arviz-plots. Is it be alright if I raised an issue and started working on it?

I know its not an appropriate time to ask, I completely understand and can revisit this next week. Thank you for your guidance sir.

**Comment by aloctavodia** [comment]
Making mistakes is the way to learn.  Yes, you can start working on a function to plot the BF in arviz-plots.

1. The next step is to check this implementation works. I added changes without actually checking the code.
2. The way this is currently implemented (after the check/fix) should work. But I wonder if this is the best way to implement it. This is a little bit hacky. I am talking about instantiating the `_DensityBase` class. Some alternatives could be:
* call the `kde` accessor to compute the densities `posterior_pdf`and `prior_pdf` and manually create the grid `posterior_grid` and `prior_grid`.
* Add one option to the `kde`  accessor to also return the grid.
* Add a new accessor that computes the kde and grid. 

Because I am unsure about the best approach here, I requested an opinion from other devs.

**Comment by PiyushPanwarFST** [comment]
Sir as you gave me the 3 suggestions although I am working on them but this could be our alternative for avoiding the instance thing from the code and just importing the density calculating functionality from arviz, it also passes all the test cases for bayes_factor



import arviz as azb
import numpy as np 
import warnings

def bayes_factor(idata, var_name, ref_val=0, return_ref_vals=False, prior= None):

    """
    Approximated Bayes Factor for comparing hypotheses of two nested models, 
    using KDE for density estimation.
    Parameters
    ----------
    idata : InferenceData
        Object containing posterior and prior data.
    var_name : str
        Name of the variable to test.
    ref_val : int or float, default 0
        Reference (point-null) value for Bayes factor estimation.
    return_ref_vals : bool, default False
        If True, also return the values of prior and posterior densities at the reference value.
    Returns
    -------
    dict
        A dictionary with Bayes Factor values: BF10 (H1/H0 ratio) and BF01 (H0/H1 ratio).
    """

    posterior = idata.posterior[var_name].values.flatten()
    prior = idata.prior[var_name].values.flatten()

    posterior = posterior[np.isfinite(posterior)]
    prior = prior[np.isfinite(prior)]

    if not isinstance(ref_val, (int, float)):
        raise ValueError("The reference value (ref_val) must be a numerical value (int or float).")

    if ref_val > posterior.max() or ref_val < posterior.min():
        warnings.warn(
            "The reference value is outside the posterior range. "
            "This results in infinite support for H1, which may overstate evidence."
        )

    posterior_grid, posterior_pdf= azb.kde(x=posterior, grid_len=512, circular=False)
    prior_grid, prior_pdf = azb.kde(x=prior, grid_len=512, circular=False)


    posterior_at_ref_val = np.interp(ref_val, posterior_grid, posterior_pdf)
    prior_at_ref_val = np.interp(ref_val, prior_grid, prior_pdf)

    bf_10 = prior_at_ref_val / posterior_at_ref_val
    bf = {"BF10": bf_10, "BF01": 1 / bf_10}

    if return_ref_vals:
        return (bf, {"prior": prior_at_ref_val, "posterior": posterior_at_ref_val})
    else:
        return bf
        
        



Q. In the 1st approach you have mentioned the kde accessor to compute the densities so is it in different file called accessors.py and or we just make the grid functionality manually and kde density remain as it is ??

**Comment by aloctavodia** [comment]
I am not sure about the purpose of the code you posted, please only post short snippets or add code to commits. If there is no `kde` function, then `azb.kde` will not work. Please check what accessors are, and how are they used, check for instance how a `kde` accessors is used here https://github.com/arviz-devs/arviz-plots/blob/main/src/arviz_plots/plots/distplot.py

The 3 options I described were not necessarily mentioned for you to explore, you don't need to solve that at this point. I was just trying to keep ideas ordered. 

I am not sure I understand your question. "
Q. In the 1st approach you have mentioned the kde accessor to compute the densities so is it in different file called accessors.py and or we just make the grid functionality manually and kde density remain as it is ??" could you please reformulate?

**Comment by PiyushPanwarFST** [comment]
For this PR, we will wait for confirmation from [Oriol Abril](https://github.com/OriolAbril) sir before proceeding further.
And for as of now can I start working on the plotting functionality of the bayes_factor function in the arviz-plots repository?

**Comment by aloctavodia** [comment]
No need to tag people again. I already said yes to that question.

File: src/arviz_stats/bayes_factor.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
+from .base.density import _DensityBase 
+import numpy as np 
+import warnings
```
Comment (aloctavodia) [inline src/arviz_stats/bayes_factor.py:3]:
```suggestion
import numpy as np 
import warnings
from arviz_base import extract

from arviz_stats.base.density import _DensityBase 
```

File: tests/test_bayes_factor.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
+import pytest
+import numpy as np
+import arviz as az
```
Comment (aloctavodia) [inline tests/test_bayes_factor.py:3]: Don't use arviz! you need to  arviz-base, stats or plots. 
```suggestion
from arviz_base import from_dict
```

File: tests/test_bayes_factor.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+from src.arviz_stats.bayes_factor import bayes_factor
+
+def test_bayes_factor_comparison():
+    idata = az.from_dict(
+        posterior={"a": np.random.normal(1, 0.5, 5000)},
+        prior={"a": np.random.normal(0, 1, 5000)}
+    )
```
Comment (aloctavodia) [inline tests/test_bayes_factor.py:11]:
```suggestion
    idata = from_dict({"posterior": {"a": np.random.normal(1, 0.5, (2, 1000))},
                        "prior": {"a": np.random.normal(0, 1, (2, 1000))}})
```

change all other call to `from_dict` following this pattern

**Review by aloctavodia — CHANGES_REQUESTED** [review]

**Review by aloctavodia — CHANGES_REQUESTED** [review]
before adding your own changes remember to pull the changes I added

File: tests/test_bayes_factor.py [hunk 12]
```diff
@@ -1,96 +1,67 @@
...
 def test_bayes_factor_custom_prior():
     posterior_data = np.random.normal(1, 0.5, 5000)
     prior_data = np.random.normal(0, 1, 5000)
     custom_prior = np.random.normal(0, 10, 5000)
-
-    idata = az.from_dict(
-        posterior={"a": posterior_data},
-        prior={"a": prior_data}
+    idata = from_dict(
+        {"posterior": {"a": posterior_data[:, np.newaxis]},
```
Comment (aloctavodia) [inline tests/test_bayes_factor.py:34]: Instead of using newaxis, pass a tuple to the shape of the np.random function

File: tests/test_bayes_factor.py [hunk 13]
```diff
@@ -27,12 +27,12 @@ def test_bayes_factor_invalid_ref_val():
         bayes_factor(idata=idata, var_name="a", ref_val="invalid")
 
 def test_bayes_factor_custom_prior():
-    posterior_data = np.random.normal(1, 0.5, 5000)
-    prior_data = np.random.normal(0, 1, 5000)
-    custom_prior = np.random.normal(0, 10, 5000)
+    posterior_data = np.random.normal(1, 0.5, (5000, 1))
```
Comment (aloctavodia) [inline tests/test_bayes_factor.py:30]: use (2, 1000)

Comment (PiyushPanwarFST) [reply tests/test_bayes_factor.py]: do we need to do it only at one place, as you mentioned, or should we do it in other test cases as well

Comment (aloctavodia) [reply tests/test_bayes_factor.py]: Everywhere

**Comment by PiyushPanwarFST** [comment]
In the arviz-plots repo, I have successfully implemented the plotting functionality for bayes_factor. All the test cases have passed, and I am able to draw all the plots properly, as you can see in the 1st screenshot. For the computation of bayes_factor, I locally installed the arviz-stats repo and imported the bayes_factor functionality into the arviz-plots repo. The arviz-plot repo, working perfectly without any issues.

Screenshot 2025-01-23 at 6 53 10 PM

If there are no conflicts or errors related to this PR (#45), could you please merge it? That way, I can directly import the bayes_factor computation functionality from arviz-stats without the need to install it locally.

I have also successfully passed all the test cases in the arviz-plots repo, as shown in the second screenshot.

Screenshot 2025-01-24 at 3 33 12 PM

Let me know if anything I need to change or improve so we can finally complete the separation of plotting and computation functionality into their respective repositories.

**Comment by aloctavodia** [comment]
Please don't comment here something related to a different repo. Instead comment on arviz-plots issue or PR and cross-reference this issue.

And again don't be pushy. You can do a PR in arviz-plot even if this pr is unmerged. Just cross-reference PRs

**Review by aloctavodia — CHANGES_REQUESTED** [review]
Ruff is failing

**Comment by PiyushPanwarFST** [comment]
Sir, I am failing the test case for `docs/readthedocs.org:arviz-stats`. To resolve this, do I need to create a `.readthedocs.yaml`  file in the root directory of the repository to configure the path for Sphinx ?

**Comment by aloctavodia** [comment]
Tests are still failing. Please make sure to run pre-commit locally, and fix errors as needed.

**Comment by PiyushPanwarFST** [comment]
Sir as per your suggestion, I run the pre-commit checks locally and noticed that pylint was failing. I resolved the 4-5 errors identified during the process. Please let me know if something needs to do or further actions are required.

File: src/arviz_stats/bayes_factor.py [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+    ref_val : int or float, default 0
+        Reference (point-null) value for Bayes factor estimation.
+    return_ref_vals : bool, default False
+        If True, also return the values of prior and posterior densities at the reference value.
+
+    Returns
+    -------
+    dict
+        A dictionary with Bayes Factor values: BF10 (H1/H0 ratio) and BF01 (H0/H1 ratio).
+    """
```
Comment (aloctavodia) [inline src/arviz_stats/bayes_factor.py:34]:
```suggestion

    References
    ---------------
    .. [1] Heck, D., 2019. A caveat on the Savage-Dickey density ratio:
       The case of computing Bayes factors for regression parameters.

    Examples
    -------------
    Moderate evidence indicating that the parameter "a" is different from zero.

    .. ipython::

        In [1]: import numpy as np
           ...: from arviz_base import from_dict
           ...: import arviz_stats as azs
           ...: idata = from_dict({"posterior":{"a":np.random.normal(1, 0.5, (2, 1000))}, 
           ...:                    {"prior":{"a":np.random.normal(0, 1, (2, 1000))}}})
           ...: azs.plot_bf(idata, var_name="a", ref_val=0)
    """
```

File: src/arviz_stats/bayes_factor.py [hunk 15]
```diff
@@ -0,0 +1,70 @@
+"""
+Module containing the Bayes Factor calculation for nested models.
+
+This module provides a function for approximating Bayes Factors using KDE for density estimation.
+"""
```
Comment (aloctavodia) [inline src/arviz_stats/bayes_factor.py:5]:
```suggestion
"""Bayes Factor using Savage-Dickey density ratio"""
```

**Review by aloctavodia — CHANGES_REQUESTED** [review]
I  made a few suggestions for the docstring. After adding them, we can merge.

**Comment by aloctavodia** [comment]
One more thing, after incorporating my suggestions, please do a rebase.
